### PR TITLE
I like to move it, move it.

### PR DESCRIPTION
**Pull request by mzhurovich**
@dkorolev Didn't manage to do a lot today :-/

### CODE REVIEW

File: Reflection/primitive_types.dsl.h [hunk 1]
```diff
@@ -1,7 +1,29 @@
+// This file is used for mass registering of primitives type handlers in reflection and serialization routines.
+// Typical usecase:
+//    #define CURRENT_DECLARE_PRIMITIVE_TYPE(typeid_index, cpp_type, current_type) \
```
Comment (dkorolev) [inline Reflection/primitive_types.dsl.h]: :+1: 

Don't see a good reason to not put the license header into this file too.

File: Reflection/reflection.h [hunk 2]
```diff
@@ -48,28 +73,29 @@ struct ReflectorImpl {
...
-      return std::string(" : ") + SuperType<T>::template CURRENT_REFLECTION_HELPER<SuperType<T>>::name();
-    }
-
-    template <typename T>
     typename std::enable_if<std::is_base_of<CurrentSuper, T>::value, std::unique_ptr<ReflectedTypeImpl>>::type
     operator()(TypeSelector<T>) {
       auto s = make_unique<ReflectedType_Struct>();
       s->name = T::template CURRENT_REFLECTION_HELPER<T>::name();
-      s->super_name = StructInheritanceString<T>();
+      s->super_name = SuperNameIfNotCurrentSuper<T>();
```
Comment (dkorolev) [inline Reflection/reflection.h]: Actually, we'll need to return not just `struct Foo {` but `struct Foo : CurrentSuper {` in C++ reflected types.

File: Reflection/test.cc [hunk 3]
```diff
@@ -162,294 +168,78 @@ TEST(Reflection, CurrentStructInternals) {
...
 
-CURRENT_STRUCT(Serializable) {
-  CURRENT_FIELD(i, uint64_t);
-  CURRENT_FIELD(s, std::string);
+// TODO: move these asserts into sources?
+static_assert(sizeof(float) == 4u, "Only 32-bit `float` is supported.");
+static_assert(sizeof(double) == 8u, "Only 64-bit `double` is supported.");
+
+CURRENT_STRUCT(StructWithAllSupportedTypes) {
+  // Integral.
```
Comment (dkorolev) [inline Reflection/test.cc]: :+1:

File: Reflection/types.h [hunk 4]
```diff
@@ -56,13 +82,15 @@ typedef std::vector<std::pair<ReflectedTypeImpl*, std::string>> StructFieldsVect
...
+  // If struct is derived from another Current struct, `super_name` contains the base type name.
+  // Empty otherwise.
   std::string super_name;
   StructFieldsVector fields;
 
   std::string CppType() override { return name; }
 
   std::string CppDeclaration() override {
-    std::string result = "struct " + name + super_name + " {\n";
+    std::string result = "struct " + name + (super_name.empty() ? "" : " : " + super_name) + " {\n";
```
Comment (dkorolev) [inline Reflection/types.h]: Yes, here, the `CurrentSuper` base in output C++ definitions of "top-level" types.

File: Reflection/primitive_types.dsl.h (hunk 1 above)
Comment (dkorolev) [inline Reflection/primitive_types.dsl.h]:
```
g++ -std=c++11 -Wall -W -ftemplate-depth=4096 -o .current/test test.cc -pthread
In file included from types.h:50:0,
                 from reflection.h:33,
                 from test.cc:28:
primitive_types.dsl.h:3:1: warning: multi-line comment [-Wcomment]
 //    #define CURRENT_DECLARE_PRIMITIVE_TYPE(typeid_index, cpp_type, current_type) \
 ^
```

File: Reflection/test.cc [hunk 5]
```diff
@@ -162,294 +168,78 @@ TEST(Reflection, CurrentStructInternals) {
...
+  CURRENT_FIELD(uint64, uint64_t, UINT64_MAX);
+  CURRENT_FIELD(int8, int8_t, INT8_MIN);
+  CURRENT_FIELD(int16, int16_t, INT16_MIN);
+  CURRENT_FIELD(int32, int32_t, INT32_MIN);
+  CURRENT_FIELD(int64, int64_t, INT64_MIN);
+  // Floating point.
+  CURRENT_FIELD(flt, float, 1e38);
+  CURRENT_FIELD(dbl, double, 1e308);
+  // Other primitive types.
+  CURRENT_FIELD(s, std::string, "The String");
```
Comment (dkorolev) [inline Reflection/test.cc]: Auch, something is broken here:

```
test.cc:191:33: error: cannot initialize a member subobject of type '::current::reflection::Field<CountFields, std::string>'
      (aka 'long long') with an lvalue of type 'const char [11]'
  CURRENT_FIELD(s, std::string, "The String");
                                ^~~~~~~~~~~~
./struct.h:129:92: note: expanded from macro 'CURRENT_FIELD'
  CURRENT_FIELD_SWITCH(__VA_ARGS__, CURRENT_FIELD_WITH_VALUE, CURRENT_FIELD_WITH_NO_VALUE)(__VA_ARGS__)
                                                                                           ^
./struct.h:136:65: note: expanded from macro 'CURRENT_FIELD_WITH_VALUE'
  ::current::reflection::Field<INSTANTIATION_TYPE, type> name = value; \
```

File: Reflection/primitive_types.dsl.h (hunk 1 above)
Comment (mzhurovich) [inline Reflection/primitive_types.dsl.h]: :+1: 
Fixed in 10061845d4233dbe79785ab43f00208b807b5575

**Comment by dkorolev** [comment]
Works in `clang++`, still does not in `g++`:

```
dima@dima-i9: ~/github/mzhurovich/Current/Reflection(reflect_and_serialize) $ make
mkdir -p .current
g++ -std=c++11 -Wall -W -ftemplate-depth=4096 -o .current/test test.cc -pthread
In file included from reflection.h:32:0,
                 from test.cc:28:
test.cc: In constructor ‘constexpr reflection_test::CURRENT_STRUCT_IMPL_StructWithAllSupportedTypes<current::reflection::DeclareFields>::CURRENT_STRUCT_IMPL_StructWithAllSupportedTypes()’:
struct.h:115:10: error: conversion from ‘const char [11]’ to non-scalar type ‘current::reflection::Field<current::reflection::DeclareFields, std::basic_string<char> > {aka std::basic_string<char>}’ requested
   struct CURRENT_STRUCT_IMPL_##s                                  \
          ^
struct.h:94:47: note: in expansion of macro ‘CURRENT_STRUCT_NOT_DERIVED’
 #define CURRENT_STRUCT_SWITCH(_1, _2, F, ...) F
                                               ^
struct.h:96:3: note: in expansion of macro ‘CURRENT_STRUCT_SWITCH’
   CURRENT_STRUCT_SWITCH(__VA_ARGS__, CURRENT_STRUCT_DERIVED, CURRENT_STRUCT_NOT_DERIVED)(__VA_ARGS__)
   ^
test.cc:175:1: note: in expansion of macro ‘CURRENT_STRUCT’
 CURRENT_STRUCT(StructWithAllSupportedTypes) {
 ^
test.cc: In member function ‘virtual void Reflection_VisitAllFields_Test::TestBody()’:
test.cc:231:31: note: synthesized method ‘constexpr reflection_test::CURRENT_STRUCT_IMPL_StructWithAllSupportedTypes<current::reflection::DeclareFields>::CURRENT_STRUCT_IMPL_StructWithAllSupportedTypes()’ first required here 
   StructWithAllSupportedTypes all;
                               ^
make: *** [.current/test] Error 1
dima@dima-i9: ~/github/mzhurovich/Current/Reflection(reflect_and_serialize) $ 
```

File: Reflection/test.cc (hunk 5 above)
Comment (mzhurovich) [inline Reflection/test.cc]: Fixed in 10061845d4233dbe79785ab43f00208b807b5575

**Comment by mzhurovich** [comment]
@dkorolev How about now?

**Comment by dkorolev** [comment]
@mzhurovich YESSSSSSSS!

**Pull request merged**